In [1462]:
import pandas as pd
from sklearn.model_selection import train_test_split,  GridSearchCV
from sklearn.preprocessing import StandardScaler , MinMaxScaler , OneHotEncoder, PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression , Ridge , Lasso
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score, mean_absolute_percentage_error

df = pd.read_csv('ToyotaCorolla.csv')

## Датасет:

In [1463]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 39 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Id                 1436 non-null   int64
 1   Model              1436 non-null   str  
 2   Price              1436 non-null   int64
 3   Age_08_04          1436 non-null   int64
 4   Mfg_Month          1436 non-null   int64
 5   Mfg_Year           1436 non-null   int64
 6   KM                 1436 non-null   int64
 7   Fuel_Type          1436 non-null   str  
 8   HP                 1436 non-null   int64
 9   Met_Color          1436 non-null   int64
 10  Color              1436 non-null   str  
 11  Automatic          1436 non-null   int64
 12  CC                 1436 non-null   int64
 13  Doors              1436 non-null   int64
 14  Cylinders          1436 non-null   int64
 15  Gears              1436 non-null   int64
 16  Quarterly_Tax      1436 non-null   int64
 17  Weight             1436 n

In [1464]:
df.describe()

,Id,Price,Age_08_04,Mfg_Month,Mfg_Year,KM,HP,Met_Color,Automatic,CC,Doors,Cylinders,Gears,Quarterly_Tax,Weight,Mfr_Guarantee,BOVAG_Guarantee,Guarantee_Period,ABS,Airbag_1,Airbag_2,Airco,Automatic_airco,Boardcomputer,CD_Player,Central_Lock,Powered_Windows,Power_Steering,Radio,Mistlamps,Sport_Model,Backseat_Divider,Metallic_Rim,Radio_cassette,Parking_Assistant,Tow_Bar
count,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.00000,1436.000000,1436.0,1436.000000,1436.000000,1436.00000,1436.000000,1436.000000,1436.000000,1436.00000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000
mean,721.555014,10730.824513,55.947075,5.548747,1999.625348,68533.259749,101.502089,0.674791,0.055710,1576.85585,4.033426,4.0,5.026462,87.122563,1072.45961,0.409471,0.895543,3.815460,0.81337,0.970752,0.722841,0.508357,0.056407,0.294568,0.218663,0.580084,0.561978,0.977716,0.146240,0.256964,0.300139,0.770195,0.204735,0.145543,0.002786,0.277855
std,416.476890,3626.964585,18.599988,3.354085,1.540722,37506.448872,14.981080,0.468616,0.229441,424.38677,0.952677,0.0,0.188510,41.128611,52.64112,0.491907,0.305959,3.011025,0.38975,0.168559,0.447751,0.500104,0.230786,0.456007,0.413483,0.493717,0.496317,0.147657,0.353469,0.437111,0.458478,0.420854,0.403649,0.352770,0.052723,0.448098
min,1.000000,4350.000000,1.000000,1.000000,1998.000000,1.000000,69.000000,0.000000,0.000000,1300.00000,2.000000,4.0,3.000000,19.000000,1000.00000,0.000000,0.000000,3.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,361.750000,8450.000000,44.000000,3.000000,1998.000000,43000.000000,90.000000,0.000000,0.000000,1400.00000,3.000000,4.0,5.000000,69.000000,1040.00000,0.000000,1.000000,3.000000,1.00000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
50%,721.500000,9900.000000,61.000000,5.000000,1999.000000,63389.500000,110.000000,1.000000,0.000000,1600.00000,4.000000,4.0,5.000000,85.000000,1070.00000,0.000000,1.000000,3.000000,1.00000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
75%,1081.250000,11950.000000,70.000000,8.000000,2001.000000,87020.750000,110.000000,1.000000,0.000000,1600.00000,5.000000,4.0,5.000000,85.000000,1085.00000,1.000000,1.000000,3.000000,1.00000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000,1.000000,1.000000,1.000000,0.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,1.000000
max,1442.000000,32500.000000,80.000000,12.000000,2004.000000,243000.000000,192.000000,1.000000,1.000000,16000.00000,5.000000,4.0,6.000000,283.000000,1615.00000,1.000000,1.000000,36.000000,1.00000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000


In [1465]:
df.describe(include='str')

,Model,Fuel_Type,Color
count,1436,1436,1436
unique,319,3,10
top,TOYOTA Corolla 1.6 16V HATCHB LINEA TERRA 2/3-...,Petrol,Grey
freq,109,1264,301


In [1466]:
NaN = df.isnull().sum().value_counts()
print(NaN)

0    39
Name: count, dtype: int64


## Предобработка: удаление скоррелированых/не нужных для поставленной задачи данных

In [1467]:
# корелляция между циллиндрами и объемом

print(df['Cylinders'].corr(df['CC']))

nan


/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/numpy/lib/_function_base_impl.py:3023: RuntimeWarning: invalid value encountered in divide
  c /= stddev[:, None]
/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/numpy/lib/_function_base_impl.py:3024: RuntimeWarning: invalid value encountered in divide
  c /= stddev[None, :]


1) Cylinders содержит единственное значение 4, значит удаляем

In [1468]:
# корелляция между передачами и объемом

print(df['Gears'].corr(df['CC']))

0.014629352089597158


### Итак, ненужными/бесполезными я счел следующие признаки: 
 - Cylinders - единственное значение
 - Weight - несущественный показатель
 - Backseat_Divider - не существенный показатель
 - Tow_Bar - несущественный показатель
 - Radio_cassette - несущественный (устаревший) показатель
 - CD_Player - несущественный (устаревший) показатель
 - Met_Color - несущественный (устаревший) показатель
 - Metallic_Rim - несущественный (устаревший) показатель
 - Mistlamps - несущественный (устаревший) показатель
 - Model - содержит слишком много значений

In [1469]:

cols_to_drop = [
    'Cylinders',
    'Weight',
    'Backseat_Divider',
    'Tow_Bar',
    'Radio_cassette',
    'CD_Player',
    'Met_Color',
    'Metallic_Rim',
    'Mistlamps',
    'Model'
]

df = df.drop(columns=cols_to_drop)

## Feature Ingeneering

### Создам 1 колонку с возрастом вместо 3 разных

In [1470]:
base_year = 2004
base_month = 8
df['Age_Month'] = (base_year*12 + base_month) - df['Mfg_Year']*12 + df['Mfg_Month'] + df['Age_08_04']

df = df.drop(columns=['Mfg_Year', 'Mfg_Month', 'Age_08_04'])

In [1471]:
print(df['Age_Month'].corr(df['KM']))

0.5049744502001694


### One-hot encoding для оставшихся str признаков

In [1472]:
df = pd.get_dummies(df, columns=['Fuel_Type'], prefix='Fuel', drop_first=True)

In [1473]:
print(df['Color'].value_counts())

Color
Grey      301
Blue      283
Red       278
Green     220
Black     191
Silver    122
White      31
Violet      4
Yellow      3
Beige       3
Name: count, dtype: int64


In [1474]:
df['Color'] = df['Color'].apply(lambda x: 'Other' if x in ['Violet', 'Yellow', 'Beige'] else x)

df = pd.get_dummies(df, columns=['Color'], prefix='Color')
df = df.drop(columns='Color_Other')

In [1475]:
df.info()
df.describe(include=['int64', 'bool'])

<class 'pandas.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 34 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Id                 1436 non-null   int64
 1   Price              1436 non-null   int64
 2   KM                 1436 non-null   int64
 3   HP                 1436 non-null   int64
 4   Automatic          1436 non-null   int64
 5   CC                 1436 non-null   int64
 6   Doors              1436 non-null   int64
 7   Gears              1436 non-null   int64
 8   Quarterly_Tax      1436 non-null   int64
 9   Mfr_Guarantee      1436 non-null   int64
 10  BOVAG_Guarantee    1436 non-null   int64
 11  Guarantee_Period   1436 non-null   int64
 12  ABS                1436 non-null   int64
 13  Airbag_1           1436 non-null   int64
 14  Airbag_2           1436 non-null   int64
 15  Airco              1436 non-null   int64
 16  Automatic_airco    1436 non-null   int64
 17  Boardcomputer      1436 n

,Id,Price,KM,HP,Automatic,CC,Doors,Gears,Quarterly_Tax,Mfr_Guarantee,BOVAG_Guarantee,Guarantee_Period,ABS,Airbag_1,Airbag_2,Airco,Automatic_airco,Boardcomputer,Central_Lock,Powered_Windows,Power_Steering,Radio,Sport_Model,Parking_Assistant,Age_Month,Fuel_Diesel,Fuel_Petrol,Color_Black,Color_Blue,Color_Green,Color_Grey,Color_Red,Color_Silver,Color_White
count,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.00000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.00000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436.000000,1436,1436,1436,1436,1436,1436,1436,1436,1436
unique,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2,2,2,2,2,2,2,2,2
top,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,False,True,False,False,False,False,False,False,False
freq,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1281,1264,1245,1153,1216,1135,1158,1314,1405
mean,721.555014,10730.824513,68533.259749,101.502089,0.055710,1576.85585,4.033426,5.026462,87.122563,0.409471,0.895543,3.815460,0.81337,0.970752,0.722841,0.508357,0.056407,0.294568,0.580084,0.561978,0.977716,0.146240,0.300139,0.002786,121.991643,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,416.476890,3626.964585,37506.448872,14.981080,0.229441,424.38677,0.952677,0.188510,41.128611,0.491907,0.305959,3.011025,0.38975,0.168559,0.447751,0.500104,0.230786,0.456007,0.493717,0.496317,0.147657,0.353469,0.458478,0.052723,36.977325,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,1.000000,4350.000000,1.000000,69.000000,0.000000,1300.00000,2.000000,3.000000,19.000000,0.000000,0.000000,3.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,17.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,361.750000,8450.000000,43000.000000,90.000000,0.000000,1400.00000,3.000000,5.000000,69.000000,0.000000,1.000000,3.000000,1.00000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,89.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,721.500000,9900.000000,63389.500000,110.000000,0.000000,1600.00000,4.000000,5.000000,85.000000,0.000000,1.000000,3.000000,1.00000,1.000000,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,137.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,1081.250000,11950.000000,87020.750000,110.000000,0.000000,1600.00000,5.000000,5.000000,85.000000,1.000000,1.000000,3.000000,1.00000,1.000000,1.000000,1.000000,0.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000,161.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## Разделение выборки данных

In [1476]:
df_X = df.drop(columns='Price')
df_Y = df['Price']

x_train, x_test, y_train, y_test = train_test_split(df_X, df_Y, test_size=0.3, random_state=42)

### Я решил разделить выборку в соотношении 70/30, ибо данных для обучения достаточно, и есть смысл оставить побольше для проверки.

### разделение данных на такие выборки нужно, ибо модель так или иначе подстраивается под данные, на которых она учится, поэтому проверка ее качества на данных, которых она училась необъективна и может привести к переобучению

## Обучение модели

In [1477]:
scaler = StandardScaler()

x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [1478]:
model = LinearRegression()

model.fit(x_train, y_train)

y_pred_train = model.predict(x_train)
y_pred = model.predict(x_test)


mse_train = mean_squared_error(y_train, y_pred_train)
r2_train = r2_score(y_train, y_pred_train)
mape_train = mean_absolute_percentage_error(y_train, y_pred_train)

mse_test = mean_squared_error(y_test, y_pred)
r2_test = r2_score(y_test, y_pred)
mape_test = mean_absolute_percentage_error(y_test, y_pred)

print(f'MSE train: {mse_train}')
print(f'R^2 train: {r2_train}')
print(f'MAPE train: {mape_train}')

print(f'MSE test: {mse_test}')
print(f'R^2 test: {r2_test}')
print(f'MAPE test: {mape_test}')

MSE train: 1224078.6657567555
R^2 train: 0.9044638542804233
MAPE train: 0.08287717973147422
MSE test: 1313850.4457755554
R^2 test: 0.905628708603394
MAPE test: 0.08524833847703835


#### вроде нормально и так, но я попробую, что будет для регуляризации

In [1479]:
ridge = Ridge()

ridge.fit(x_train, y_train)

y_pred_train = ridge.predict(x_train)
y_pred = ridge.predict(x_test)


mse_train = mean_squared_error(y_train, y_pred_train)
r2_train = r2_score(y_train, y_pred_train)
mape_train = mean_absolute_percentage_error(y_train, y_pred_train)

mse_test = mean_squared_error(y_test, y_pred)
r2_test = r2_score(y_test, y_pred)
mape_test = mean_absolute_percentage_error(y_test, y_pred)

print(f'MSE train: {mse_train}')
print(f'R^2 train: {r2_train}')
print(f'MAPE train: {mape_train}')

print(f'MSE test: {mse_test}')
print(f'R^2 test: {r2_test}')
print(f'MAPE test: {mape_test}')

MSE train: 1224295.0461629217
R^2 train: 0.9044469663543507
MAPE train: 0.08283351695217811
MSE test: 1315804.0266760157
R^2 test: 0.9054883867326536
MAPE test: 0.08521894793570403


In [1480]:
lasso = Lasso()

lasso.fit(x_train, y_train)

y_pred_train = lasso.predict(x_train)
y_pred = lasso.predict(x_test)


mse_train = mean_squared_error(y_train, y_pred_train)
r2_train = r2_score(y_train, y_pred_train)
mape_train = mean_absolute_percentage_error(y_train, y_pred_train)

mse_test = mean_squared_error(y_test, y_pred)
r2_test = r2_score(y_test, y_pred)
mape_test = mean_absolute_percentage_error(y_test, y_pred)

print(f'MSE train: {mse_train}')
print(f'R^2 train: {r2_train}')
print(f'MAPE train: {mape_train}')

print(f'MSE test: {mse_test}')
print(f'R^2 test: {r2_test}')
print(f'MAPE test: {mape_test}')

MSE train: 1224252.5789315153
R^2 train: 0.9044502808109469
MAPE train: 0.08284630291381033
MSE test: 1316556.2775436996
R^2 test: 0.9054343540335243
MAPE test: 0.08525450135678891


In [1481]:
# перед этим я дропну еще признаков, ибо PolynomialFeatures даже для deg = 2 добавит еще С(34, 2) = 34*33/(2!) признаков
# spoiler: это ухудшило качество, поэтому я попробую приемнить для текущего датасета

steps = [
    ('poly', PolynomialFeatures(degree=2)),
    ('model', Lasso(max_iter=100))
]
pipe = Pipeline(steps)

# + небольшой поиск руками
param_grid = {'model__alpha': [0.01, 0.1, 1.0, 30.0, 100.0]}
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='r2')
grid.fit(x_train, y_train)

best = grid.best_estimator_
print(f"Best param: {grid.best_params_['model__alpha']}")

y_pred_lasso_train = best.predict(x_train)
y_pred_lasso = best.predict(x_test)

r2_train = r2_score(y_train, y_pred_lasso_train)
mape_lasso_train = mean_absolute_percentage_error(y_train, y_pred_lasso_train)
r2 = r2_score(y_test, y_pred_lasso)
mape_lasso = mean_absolute_percentage_error(y_test, y_pred_lasso)

print(f"R^2 train: {r2_train}")
print(f"R^2 test: {r2}")
print(f"MAPE train: {mape_lasso_train}")
print(f"MAPE test: {mape_lasso}")

/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.625e+08, tolerance: 1.083e+06
  model = cd_fast.enet_coordinate_descent(
/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.477e+08, tolerance: 9.757e+05
  model = cd_fast.enet_coordinate_descent(
/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check t

Best param: 30.0
R^2 train: 0.9379350082098635
R^2 test: 0.9227168994414096
MAPE train: 0.06770540657248877
MAPE test: 0.08091529605517175


/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.865e+07, tolerance: 1.004e+06
  model = cd_fast.enet_coordinate_descent(
/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.142e+07, tolerance: 1.059e+06
  model = cd_fast.enet_coordinate_descent(
/Users/alex/Desktop/ml-course-homeworks/.venv/lib/python3.13/site-packages/sklearn/linear_model/_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check t

In [ ]:
# и для ridge

steps = [
    ('poly', PolynomialFeatures(degree=2)),
    ('model', Ridge(max_iter=100))
]
pipe = Pipeline(steps)

# + небольшой поиск руками
param_grid = {'model__alpha': [0.01, 0.1, 1.0, 30.0, 100.0]}
grid = GridSearchCV(pipe, param_grid, cv=5, scoring='r2')
grid.fit(x_train, y_train)

best = grid.best_estimator_
print(f"Best param: {grid.best_params_['model__alpha']}")

y_pred_lasso_train = best.predict(x_train)
y_pred_lasso = best.predict(x_test)

r2_train = r2_score(y_train, y_pred_lasso_train)
mape_lasso_train = mean_absolute_percentage_error(y_train, y_pred_lasso_train)
r2 = r2_score(y_test, y_pred_lasso)
mape_lasso = mean_absolute_percentage_error(y_test, y_pred_lasso)

print(f"R^2 train: {r2_train}")
print(f"R^2 test: {r2}")
print(f"MAPE train: {mape_lasso_train}")
print(f"MAPE test: {mape_lasso}")

Best param: 30.0
R^2 train: 0.9604599407552212
R^2 test: 0.8754593127749503
MAPE train: 0.05345756283980768
MAPE test: 0.09627228461440855


- **Данных немного, поэтому и скорость обучения высокая**
- **Регуляризация не улучшила точность -> мультиколлинеарность отсутствует**
- **Регуляризация с полиномиальными признаками и гипепараметром для Лассо немного улучшила точность -> существует некоторая нелинейная зависимость между данными**
- **В случае Лассо с гиперпараметром предсказания для train и test примерно одинаковой точности -> модель не запомнила много шума**
- **В случае Ridge с гиперпараметром предсказания для train и test разной точность (по r2: 0.96 и 0.87 соответсвенно) -> существуют лишние/коллинеарные признаки и произошло переобучение**

## Итого

1) Использованные метрики:
    - **MSE** - для оценки суммы обьективных отклонений (не MAE, т.к. он менее чувствителен к выбросам, которые критичны для данного датасета)
    - **MAPE** - для оценки ошибки в процентах (усредненной)
    - **R^2** - для оценки дисперсии нашей цены исходя из работы модели

2) На какой выборке метрики:
    - Также я считал метрики для тестовой и тренировочной выборок, чтобы оценить влияние шума и риски переобучения

3) Лучше по итогу справилась:
    - модель Лассо с гиперпараметром l = 30 для выборки с дополнительными полиномиальными признакакми - R^2 test: 0.92, MAPE test: 0.08. Почему лассо а не ridge - как видно из ridge, модель переобучилась - значит существуют лишние признаки, которые как раз были аннулированы регуляризацией Лассо, обнуляющей лишние признаки
    - Но также неплохо справилась модель без регуляризации и полиномиальных признаков - R^2: 0.905, MAPE: 0.085 - что говорит об отсутсвии коллизий в выборке данных

4) Насколько хорошие получились результаты:
    - В целом, меня они устроили, ибо точность довольно приемлимая (для средней цены по датасету 10730.8 модель предсказывает с точностью до 8-10% ~ +- 1000 евро (вроде евро)), однако единственное, что мне не особо нравится - большое число признаков, из-за чего регуляризация с полиномиальными признаками в случае с бОльшим числом данных сильно замедлится. С другой стороны, и без нее результат довольно точный, поэтому я в целом доволен этой работой

5) Чем докажете, что ваша модель не переобучилась?
    - Во всех случаях, кроме последнего (ridge с гиперпараметром) точность предсказантя модели почти не отличается на тренировочной и тестовой выборке, что говорит о том, что она не запомнила шум и не выучила тренировочный датасат, т.е. не переобучилась.